### PR TITLE
Fix controller-manager race condition issue which cause endpoints flush during restart

### PR DESCRIPTION
**Pull request by xinxiaogang**
Fix: Wait for endpoints to become available instead of assuming there are no endpoints. (On kube-controller-manager restart.)

### CODE REVIEW

File: pkg/controller/endpoint/endpoints_controller.go [hunk 1]
```diff
@@ -305,6 +305,21 @@ func (e *EndpointController) syncService(key string) {
 		e.queue.Add(key) // Retry
 		return
 	}
+	if len(pods.Items) == 0 {
```
Comment (bprashanth) [inline pkg/controller/endpoint/endpoints_controller.go]: Can you do https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/replication/replication_controller.go#L500 instead?

File: pkg/controller/endpoint/endpoints_controller.go [hunk 2]
```diff
@@ -305,6 +305,21 @@ func (e *EndpointController) syncService(key string) {
...
+		// When controller manager starts, usually service list would be faster than pod list;
+		// So list pods from cache would get empty list;
+		// The risk is endpoints controller then will flush the endpoints for this service;
+		// If both controller manager and kube-proxy are using service VIP to access apiserver,
+		// they will fail after endpoints flush, and there is no way to automatically recover.
```
Comment (bprashanth) [inline pkg/controller/endpoint/endpoints_controller.go]: Hmm, they should keep retrying. Is that not the case? a service with no endpoints gets a reject rule in iptables, and when it gets endpoints the reject is removed and the endpoints are inserted.

**Comment by bprashanth** [comment]
@kubernetes/sig-api-machinery

File: pkg/controller/endpoint/endpoints_controller.go (hunk 1 above)
Comment (lavalamp) [inline pkg/controller/endpoint/endpoints_controller.go]: +1, that's the correct fix.

**Comment by lavalamp** [comment]
This looks like a good thing to fix, just we should fix by waiting for the pod store to sync instead of by listing every pod. I think we should consider this for a cherry-pick. (Problem statement: you restart controller-manager and all of your endpoints briefly go away.)

**Comment by lavalamp** [comment]
(Not urgent to get into 1.2.0, too late anyway. For 1.2.1.)

File: pkg/controller/endpoint/endpoints_controller.go (hunk 1 above)
Comment (xinxiaogang) [inline pkg/controller/endpoint/endpoints_controller.go]: @bprashanth Good suggestion. Updated and please review again.

File: pkg/controller/endpoint/endpoints_controller.go (hunk 2 above)
Comment (xinxiaogang) [inline pkg/controller/endpoint/endpoints_controller.go]: @bprashanth What I observed is iptables keep there to dnat outgoing packets to kube-proxy, but kube-proxy can not do any forwarding.

**Comment by bprashanth** [comment]
Thanks, LGTM

**Comment by lavalamp** [comment]
Thanks for this! I took the liberty of editing your initial comment so it'll be clear to the people evaluating the cherry-pick what this does. :)

**Pull request merged**